MM-64878: FIPS Build

lieut-data opened the pull request:
#### Summary
Start using the Chainguard container to run tests as well as build a FIPS compatible image for testing purposes.

#### Ticket Link
Fixes: https://mattermost.atlassian.net/browse/MM-64878

#### Release Note
```release-note
NONE
```

agarciamontoro reviewed (approved):
Yay, this is finally taking shape! Thank you for the thorough work here, I know that dealing with pipelines is never fun 😂 I've left a bunch of comments, all of them non-blocking, and most of them for my own understanding :)

lieut-data commented on .github/workflows/mmctl-test-template.yml:
```diff
@@ -25,8 +36,30 @@ jobs:
+      - name: Setup BUILD_IMAGE
+        id: build
+        run: |
+          if [[ ${{ inputs.fips-enabled }} == 'true' ]]; then
+            echo "IMAGE=mattermost/mattermost-build-server-fips:${{ inputs.go-version }}" >> "${GITHUB_OUTPUT}"
+            echo "LOG_ARTIFACT_NAME=${{ inputs.logsartifact }}-fips" >> "${GITHUB_OUTPUT}"
+          else
+            echo "IMAGE=mattermostdevelopment/mattermost-build-server:${{ inputs.go-version }}" >> "${GITHUB_OUTPUT}"
+            echo "LOG_ARTIFACT_NAME=${{ inputs.logsartifact }}" >> "${GITHUB_OUTPUT}"
+          fi
```
Just a short term workaround -- after chatting with @esarafianou, I think we'll end up with both of these permanently in the `mattermostdevelopment` space (and private).

lieut-data commented on .github/workflows/mmctl-test-template.yml:
```diff
@@ -44,35 +78,39 @@ jobs:
+            --env FIPS_ENABLED="${{ inputs.fips-enabled }}" \
             -v $PWD:/mattermost \
             -w /mattermost/server \
+            --entrypoint "/usr/bin/make" \
```
Now removed 🎉

lieut-data commented on server/build/docker/postgres.conf:
```diff
@@ -5,3 +5,4 @@ full_page_writes = off
 default_text_search_config = 'pg_catalog.english'
 commit_delay=1000
 logging_collector=off
+password_encryption = 'scram-sha-256'
```
Yes, we're defaulting the test Postgres configuration to be FIPS compatible (still works with non-FIPS of course).

lieut-data commented on server/build/Dockerfile.fips:
```diff
@@ -0,0 +1,50 @@
+  tzdata 
+
+# Set mattermost group/user and download Mattermost
+COPY $MM_PACKAGE /tmp/mattermost.tgz
```
Good question! I admit I copied this part over from the original changes, but I tried building this and it definitely /doesn't/ work. We also don't seem to try building either image anywhere in CI, so I'm uncertain what this is supposed to be...

... I'm inclined to track this as a separate "issue" and fix it up with another PR once I understand the broader thing. For now, I'll just delete the file since it's wrong anywway.

lieut-data commented on server/platform/shared/filestore/s3store.go:
```diff
@@ -675,14 +662,19 @@ func (b *S3FileBackend) RemoveDirectory(path string) error {
+		defer cancel2()
+		err := b.client.RemoveObject(ctx2, b.bucket, object.Key, s3.RemoveObjectOptions{})
+		if err != nil {
+			return errors.Wrapf(err, "unable to remove object %s from directory %s", object.Key, path)
```
Agreed -- I'll track restoring this for at least non-FIPS builds.

agarciamontoro commented on server/Makefile:
```diff
@@ -106,6 +106,12 @@ GOFLAGS ?= $(GOFLAGS:)
+
+FIPS_ENABLED ?= false
+ifeq ($(FIPS_ENABLED),true)
+	BUILD_TAGS += requirefips
```
> but we could also use for enforcing license validation as needed

This is what I was envisioning for license validation, something like what you came up with for service environment. I still need to figure out what to do when the license is not there but the binary is FIPS-compliant, though.

lieut-data commented on server/build/Dockerfile.fips:
```diff
@@ -0,0 +1,83 @@
+# First stage - FIPS dev image with dependencies for building
```
@esarafianou, updated this Dockerfile to use multiple stages so we don't ship with the `-dev` version of the Chainguard image.

lieut-data commented:
/update-branch

lieut-data commented:
@DHaussermann & @saturninoabril, I found what I believe to be the issue with the FIPS Docker image, fixed it in the enterprise pipeline, and added a test to keep this from regressing. 

In short, my local testing explicitly used the `Dockerfile.fips`, while the CI missed this. We still got an image, but the Go binary self checked and failed on the required OpenSSL configuration.

Note that I fully expect we'll find more gotchas with the FIPS build, especially once we have e2e tests running, but can iterate on this without impacting our confidence in the main release.

DHaussermann commented:
Ah! Thanks @lieut-data 
If I understand correctly the OpenSSL configuration. iSsue for CI is now resolved?

I'm curious what next steps are for this and if we're now ready to merge and cherry pick.

lieut-data commented:
> Ah! Thanks @lieut-data If I understand correctly the OpenSSL configuration. iSsue for CI is now resolved?
> 
> I'm curious what next steps are for this and if we're now ready to merge and cherry pick.

Yes, that's right @DHaussermann. With the testing of the non-FIPS so far, my recommendation would be to merge and cherry-pick to unblock @saturninoabril's ability to setup the e2e tests and surface any new issues then. I can do the merge once we're lifted the `3: QA Review` label.

mattermost-build commented:
Cherry pick is scheduled.
